resources/2024-06-resource-extensibility-feature-spec.md

```diff
@@ -0,0 +1,452 @@
+
+- Recipes for user defined types: I can create Recipes for the custom resource types and deploy via Radius 
+
+- OSS contributions: I can create and contribute Recipes for the custom resource types for the community to use.
+
+## Key scenarios
+
+### Scenario 1: Deb integrates the Budgets app with Radius
+
+Deb, a platform engineer at a Financial Services company, wants to integrate the Budgets app with Radius. The Budgets app relies on an internal messaging service called Plaid. Deb wants to use Plaid as a resource type in Radius to deploy the application seamlessly. He needs a way to define and use Plaid as a custom resource type in Radius without having to write Go code. This will enable him to leverage all the features of Radius, such as Recipes, Connections, and the rad CLI, with ease and flexibility.
+
+### Scenario 2: Amy contributes and open-sources PostgreSQL support to Radius
+
+Amy is a system integrator who helps customers build cloud native applications on AWS. Amy heard about Radius and starts integrating their customer workloads in Radius. She sees a lot of her scenarios involve PostgreSQL and finds Radius doesn’t support that yet. She wants to contribute the PostgreSQL support to Radius and open source it so anyone can use it
+
+### Scenario 3: Raj publishes Recipes for proprietary services
+
+Raj is a partner or a system integrator who helps enterprises integrate proprietary service for eg : Oracle database in their cloud native applications. Raj tries out Radius and wants to write Recipes to create an Oracle Database with a standard set of best practices and policies integrated. He works with a lot of customers who are wanting to use the Oracle Database recipe in their applications.
+
+## Key dependencies and risks
+<!-- What dependencies must we take in order to enable this scenario? -->
+
+<!-- What other risks are you aware of that need to be mitigated. If you have a mitigation in mind, summarize here. -->
+<!-- **Dependency Name** – summary of dependency.  Issues/concerns/risks with this dependency -->
+<!-- **Risk Name** – summary of risk.  Mitigation plan if known. If it is not yet known, no problem. --> 
+
+- Dependency: Bicep compiler merge support
+    - We need to ensure that the ongoing Bicep compiler merge work is complete so that custom resource types can be dynamically generated for the user defined resource types.
+    - Mitigation: Work closely with the Bicep team to ensure that the merge is completed on time and that the necessary features are available to enable custom resource types in Radius .
+
+- Risk: Adoption and community engagement
+    - There is a risk that users may not fully embrace the extensibility feature or contribute custom resource types to the community.
```

Comment:
We can add a few custom resource types and a tutorial to show external contributors how they might do it.

Reply:
Are we going to be requiring functional tests for all the custom resource types added? That might have some security issues in the pipelines we have. Like secrets and stuff.

Reply:
flush some more details on how we want to make this experience simple and easier to use

Reply:
Added some details around secrets handling and functional tests in the security risk below.